elytron/WFCORE-5178_native-support-for-oidc.adoc

```diff
@@ -0,0 +1,178 @@
+
+==== Token storage
+
+The new HTTP mechanism implementation will be able to make use of a session token store and a cookie token store
+to store account info (i.e., access tokens, ID tokens, and refresh tokens).
+
+The session token store will be used by default but it should be possible to specify that the cookie token store
+should be used instead.
+
+The new HTTP mechanism implementation will refresh the access token when it is expired. Since the Keycloak client adapter
+provides an option to refresh the token on every request, we should provide this option as well but make
+sure to document that the use of this option is discouraged since it can have a significant impact on performance.
+For more details about what will be documented, see the description of the `always-refresh-token` option
+https://www.keycloak.org/docs/latest/securing_apps/#_refresh_token_each_req[here].
+
+==== New elytron-oidc-client subsystem
+
+A new subsystem, `elytron-oidc-client` will be added to WildFly. Its schema will be similar to the existing
+https://github.com/keycloak/keycloak/blob/master/adapters/oidc/wildfly/wildfly-subsystem/src/main/resources/schema/wildfly-keycloak_1_1.xsd[keycloak-adapter-subsystem].
+The new subsystem will install a `DeploymentUnitProcessor` which will detect if OIDC is required for a deployment.
+
+Like with the existing Keycloak client adapter, the OIDC configuration will be specified via the deployment
+or via subsystem configuration. For the deployment case, the OIDC configuration will be specified in a JSON
+configuration file in the WEB-INF directory along with the `auth-method`, `OIDC`, specified in the `web.xml`.
+For the subsystem case, the `auth-method` won't need to be specified in the `web.xml` file, it will be set
+automatically.
+
+Once a deployment has been processed, if its `auth-method` is `OIDC`, the OIDC integration will be activated.
+Otherwise, no activation will occur and the deployment will continue to use the existing security capabilities
+provided by WildFly.
+
+If the OIDC integration is activated for a deployment, then appropriate module dependencies will be added to
+the deployment (e.g., a dependency on the new Elytron OIDC module). This will ensure the activation of the
+OIDC HTTP mechanism implementation for the deployment. The deployment will make use of a "Virtual Security Domain",
+similar to what was implemented for https://github.com/wildfly/wildfly-proposals/blob/master/microprofile/WFLY-12609_MicroProfile_JWT_1_1.adoc#hard-requirements[Microprofile JWT].
+This means that it will no longer be necessary to pre-configure SecurityDomains in the Elytron subsystem,
+as was required with the Keycloak client adapter.
+
+It should be possible to make use of different OpenID providers with WildFly. While working on the
+implementation for this RFE, we should make sure that we aren't including Keycloak-specific configuration, if
```

Comment:
Are there any Keycloak-specific configuration options we want to include? If it's the case, why do want those?

Reply:
There are some options (like "realm" and "auth-server-base-url" and some clustering-related options like "register-node-at-startup") that only make sense with the Keycloak provider. I've kept these so that it's still possible to configure things with Keycloak easily.

The comment in the document here about avoiding Keycloak-specific things was mostly just to ensure that other OpenID providers could be used in addition to Keycloak.

Reply:
To avoid confusion, what about adding a prefix for such options ("keycloak-"), or move those to a separate vendor-specific configuration (even a generic map?).

When I work with Keycloak client, it's hard for me to understand which options are relevant to me. There are a lot of them, badly documented in some cases.

Reply:
An advantage of using the same option names as the existing Keycloak adapter is that the Keycloak web console can be used to generate the JSON content or the subsystem config for the user so they can easily update their deployment or subsystem config. If we were to add a prefix to these option names or move them into a separate map, then it wouldn't be possible to use the Keycloak web console to generate the config for user (unless that is updated as well).

Instead of adding a prefix to these option names or moving them into a separate map, what if I update the description of these options to emphasize that these are Keycloak-specific options?

Reply:
In case we want to keep such functionality/compatibility, documentation and good descriptions in management model / XSD are the only way. Improvement of descriptions would help a lot for sure.